cirq-core/cirq/ops/state_preparation_gate_test.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import cirq
+import pytest
+
+
+def test_state_prep_gate():
+    states = np.array(
+        [
+            [1, 0, 0, 0],
+            [1, 0, 0, 1],
+            [3, 5, 2, 7],
+            [0.7823, 0.12323, 0.4312, 0.12321],
+            [23, 43, 12, 19],
+            [1j, 0, 0, 0],
+            [1j, 0, 0, 1j],
+            [1j, -1j, -1j, 1j],
+            [1 + 1j, 0, 0, 0],
+            [1 + 1j, 0, 1 + 1j, 0],
+        ]
+    )
+    for state in states:
+        gate = cirq.PrepareState(state)
+        qubits = cirq.LineQubit.range(2)
+        circuit = cirq.Circuit(
+            [
+                cirq.H(qubits[0]),
+                cirq.CNOT(qubits[0], qubits[1]),
+                gate(qubits[0], qubits[1]),
+            ]
+        )
+        simulator = cirq.Simulator()
```

Comment:
Is there a strong need to use the simulator in this test, as opposed to directly inspecting the decomposed form? I don't see that any other gate tests invoke the simulator, and I'd like to avoid expanding the scope of this test further than is necessary.

Reply:
I haven't really fixed this because the implementation got switched over to Kraus operators. I feel that just comparing to check if it's the right operator doesn't serve the purpose of a test, or maybe that's just be because I didn't get what the operator does intuitively. Should I still try to get rid of the simulator, and if so how, just by comparing the value of the kraus operator for small values of state?

Reply:
Comparing against the expected kraus operators is what we've done for other tests, and it's what I recommend here. See the [ResetChannel test](https://github.com/quantumlib/Cirq/blob/578c956b3a9f5d13626b58a3993c084bdb52a6d5/cirq-core/cirq/ops/common_channels_test.py#L444) for an example of this.

We prefer these small tests to make identifying breakages easier; if a change only causes `simulator_test` to fail, the problem is clear, but if it causes all of the gate tests to fail as well it becomes challenging to pin down the source of the issue. The reason we can do this is that the simulator tests provide coverage for simulation of arbitrary operators; when adding a new operator, it's sufficient to verify that it has the proper unitary or Kraus-operator representation.

Reply:
Okay, I did that, simulator removed.